File: Inheritance/2.rb
class Calc  #대문자로 시작
  #생성자
  #메소드 생성
  def initialize(v1,v2)
    #p v1, v2
    @v1 = v1  # @v1: 인스턴스 변수
    @v2 = v2
  end
  def add()
    return @v1+@v2
  end
  def sub()
    return @v1-@v2
  end
  def setV1(v)
    if v.is_a?(Integer) # v가 Integer인지?
      @v1 = v
    end
  end
  def getV1()
    return @v1
  end
end

class CalcMul < Calc
  def mul()
    return @v1*@v2
  end
end

class CalcDiv < CalcMul
  def div()
    return @v1/@v2
  end
end

c1 = CalcDiv.new(20,10)
p c1, c1.div()
#<CalcDiv:0x000000000292e670 @v1=20, @v2=10>
p c1, c1.mul()
#<CalcDiv:0x000000000292e670 @v1=20, @v2=10>
p c1, c1.add()
#<CalcDiv:0x000000000292e670 @v1=20, @v2=10>
p c1, c1.sub()
#<CalcDiv:0x000000000292e670 @v1=20, @v2=10>

File: Inheritance/2.py
class Calc(object):
    def __init__(self,v1,v2): # 첫번째 입력값이 v1, 두번째 입력값이 v2
        # print(v1, v2)
        if isinstance(v1, int):
            self.v1 = v1    #인스턴스 변수
        if isinstance(v2, int):
            self.v2 = v2
    def add(self):  #관습적으로 self라고 함
        return self.v1+self.v2
    def sub(self):
        return self.v1-self.v2
    # def mul(self):
    #     return self.v1*self.v2
    def setV1(self, v):
        if isinstance(v, int):
            self.v1 = v
    def getV1(self):
        return self.v1

class CalcMul(Calc):
    def mul(self):
        return self.v1*self.v2
c1 = CalcMul(10,10)
print(c1.mul())
print(c1.add())
print(c1.sub())

class CalcDiv(CalcMul):
    def div(self):
        return self.v1/self.v2
c2 = CalcDiv(20,10)
print(c2, c2.mul())
# <__main__.CalcDiv object at 0x0000019FBB78D748> 200
print(c2, c2.div())
# <__main__.CalcDiv object at 0x0000019FBB78D748> 2.0
